src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -573,13 +601,22 @@ void ObjectSynchronizer::exit(oop object, BasicLock* lock, JavaThread* current)
     markWord mark = object->mark();
     if (LockingMode == LM_LIGHTWEIGHT) {
       // Fast-locking does not use the 'lock' argument.
-      while (mark.is_fast_locked()) {
-        const markWord new_mark = mark.set_unlocked();
-        const markWord old_mark = mark;
-        mark = object->cas_set_mark(new_mark, old_mark);
-        if (old_mark == mark) {
-          current->lock_stack().remove(object);
-          return;
+      LockStack& lock_stack = current->lock_stack();
+      if (mark.is_fast_locked() && lock_stack.try_recursive_exit(object)) {
+        // Recursively unlocked.
+        return;
+      } else if (mark.is_fast_locked() && lock_stack.is_recursive(object)) {
+        // This lock is recursive but unstructured exit. Just inflate the lock.
```

Comment:
Again this seems in the wrong place - this should be a very rare case so we should not be checking it explicitly before the expected cases!

Reply:
In exit we must always check for recursions first. Unsure what you are proposing here.

Maybe you want to call remove first, and have a branch on if the number removed is greater than 1. And in that case inflate an update the recessions field before falling through. Something like this:
```c++
      // Fast-locking does not use the 'lock' argument.
      LockStack& lock_stack = current->lock_stack();
      if (mark.is_fast_locked()) {
        if (lock_stack.try_recursive_exit(object)) {
          // Recursively unlocked.
          return;
        }

        size_t recursions = lock_stack.remove(object) - 1;

        if (recursions == 0) {
          while (mark.is_fast_locked()) {
              const markWord new_mark = mark.set_unlocked();
              const markWord old_mark = mark;
              mark = object->cas_set_mark(new_mark, old_mark);
              if (old_mark == mark) {
                return;
              }
            }
        }

        lock_stack.push(object);
        ObjectMonitor* mon = inflate(current, object, inflate_cause_vm_internal);
        if (mon->is_owner_anonymous()) {
          mon->set_owner_from_anonymous(current);
        }
        mon->set_recursions(recursions);
      }
```

This make the code a little more like the emitted code. Except it is conditioned on the mark word lock bits.
Hard to believe this will have a measurable difference. But at least to me it is more noisy.

Reply:
It wasn't the recursions I was querying but the unstructured locking aspect.

Reply:
To be clear, my concern is that for a simple exit we not only have to first check for a recursive exit (fine) but also this unexpected rare unstructured locking recursive case. Thinking it through part of the problem is that a simple-exit does itself allow for unstructured locking. Is it worth adding an additional case to peek at the top of the lock-stack and then do an exit with a pop for the most common non-recursive case? That way we in effect handle things as follows:
- recursive exit
- direct exit
- recursive unstructured exit
- direct unstructured exit

Reply:
First of let us note that when reaching this code the unstructured exit is the common case. The normal exit and recursive exit is usually handled in the emitted code (this includes the interpreter). We reach this because either a CAS failed somewhere due to a concurrent hashCode instalment, or the exit was unstructured. Inflated monitors exit just jumps passed this code (everything is conditioned on `mark.is_fast_locked()`). 

Is this motivated by the structure/layout of the C++ code. Or an optimisation?

If it is motivated by the structure/layout. Then we can lay it out as you described. It would add some code duplication.

If it is motivated as an optimisation then after the recursive exit fail, we should just call remove and act based on the return value.

Reply:
Perhaps changing this comment will help. Please consider:
```
  // This lock is recursive but is not at the top of the lock stack so we're
  // doing an unstructured exit. We have to fall thru to inflation below and
  // let ObjectMonitor::exit() do the unlock.
```

This check on L570 has to be done before the block that begins on L572
because that block assumes that `object` is not a recursive lock and it
will remove all instances of `object` from the lock stack on L578. That
would unexpectedly unlock `object` too many times.

Just as a side note: The block from L572 -> L584 also handles both
a structured exit of a non-recursive lock AND an unstructured exit of a
non-recursive lock. This is true because we use `remove()` on L578
instead of a `pop()` which assumes top-of-stack/structured locking.